tests/core/pyspec/eth2spec/test/context.py

```diff
@@ -347,56 +359,91 @@ def decorator(fn):
     return decorator
 
 
+def _get_preset_targets(kw):
+    preset_name = DEFAULT_TEST_PRESET
+    if 'preset' in kw:
+        preset_name = kw.pop('preset')
+    return spec_targets[preset_name]
+
+
+def _get_run_phases(phases, kw):
+    """
+    Return the fork names for the base `spec` in test cases
+    """
+    if 'phase' in kw:
+        # Limit phases if one explicitly specified
+        phase = kw.pop('phase')
+        if phase not in phases:
+            dump_skipping_message(f"doesn't support this fork: {phase}")
+            return None
+        run_phases = [phase]
+    else:
+        # If pytest `--fork` flag is set, filter out the rest forks
+        run_phases = set(phases).intersection(DEFAULT_PYTEST_FORKS)
+
+    return run_phases
+
+
+def _get_available_phases(run_phases, other_phases):
+    """
+    The return the available fork names for multi-phase tests
+    """
+    available_phases = set(run_phases)
+    if other_phases is not None:
+        available_phases |= set(other_phases)
+    return available_phases
+
+
+def _run_test_case_with_phases(fn, phases, other_phases, kw, args, is_fork_transition=False):
```

Comment:
one thing that seems a little off is that we are passing information about one specific type of test, the `fork_transition` test type, into this function signature when this function should work generically across any type of test....

i think to change it would require some substantial refactoring so this may be ok for now, but i'll leave this comment to call it to our attention...

and more generally, reading all of this code makes me realize we may be starting to stretch how much we are using this test infra, e.g. there are references to phases, but also forks. and specs, but also configs and presets. it is not something to do immediately but a broader refactor may be on the horizon!

Reply:
> `is_fork_transition`

Right, I want to get rid of it too but (i) the place we do `for fork_meta in fork_metas` for-loop and (ii) the place we get `run_phases` are mixed together.

There are three cases around `run_phases`:
1. From test generator call: `kw['phase']` is assigned
2. From pytest call, without `--fork` flag: use all phases that the test cases are assigned
3. From pytest call, with `--fork` flag: use `DEFAULT_PYTEST_FORKS` would filter `phases`

The current drawback: because `--fork` filter is inside `_run_test_case_with_phases`, if I set `--fork=phase0`, it still runs `_run_test_case_with_phases` two times due to the `for fork_meta in fork_metas`. Therefore this `is_fork_transition` is here to prevent it from signaling `pytest.skip`.

I'll take a look if it can be refactored in a better way. 👀 

--

> and more generally, reading all of this code makes me realize we may be starting to stretch how much we are using this test infra, e.g. there are references to phases, but also forks. and specs, but also configs and presets. it is not something to do immediately but a broader refactor may be on the horizon!

Totally agreed! It bothered me for a long time and is worth another deep clean-up.

Reply:
yeah i'm not sure there is a better solution for now 😅 

worth keeping in mind all of the various use cases if/when we get around to a bigger refactor :)